test/end2end_test.go

```diff
@@ -3116,6 +3116,42 @@ func testMetadataUnaryRPC(t *testing.T, e env) {
 	}
 }
 
+func (s) TestMetadataOrderUnaryRPC(t *testing.T) {
+	for _, e := range listTestEnv() {
+		testMetadataOrderUnaryRPC(t, e)
+	}
+}
+
+func testMetadataOrderUnaryRPC(t *testing.T, e env) {
+	te := newTest(t, e)
+	te.startServer(&testServer{security: e.security})
+	defer te.tearDown()
+	tc := testpb.NewTestServiceClient(te.clientConn())
+
+	ctx := metadata.NewOutgoingContext(context.Background(), testMetadata)
+	ctx = metadata.AppendToOutgoingContext(ctx, "key1", "value2")
+	ctx = metadata.AppendToOutgoingContext(ctx, "key1", "value3")
+
+	newMetadata, _ := metadata.FromOutgoingContext(ctx)
```

Comment:
I'm a bit paranoid about this. `FromOutgoingContext` could also return the wrong order.

How about build the expected metadata explicitly:
```go
newMetadata := metadata.Join(testMetadata, metadata.Pairs("key1", "value2","key1", "value3"))
```

Reply:
Yes, `FromOutgoingContext` is some wrap of `Join`, we can using `Join` directly.